File: src/my_project_typescript/medium_problems_typescript/from1to50/subtreeOfAnotherTreeLogn.ts

// * Definition for a binary tree node.
 class TreeNode {
     val: number
     left: TreeNode | null
     right: TreeNode | null
     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
         this.val = (val===undefined ? 0 : val)
         this.left = (left===undefined ? null : left)
         this.right = (right===undefined ? null : right)
     }
 }


 function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
     let v1: boolean 
     let v2: boolean 
     let v3: boolean 

    if (root === null){
        return false
    }
    else {
        v1 = false 
        if (root.val === subRoot.val){
            v1 = checkSubtree(root,subRoot)
        }
        v2 = isSubtree(root.left, subRoot.left)
        v3 = isSubtree(root.right, subRoot.right)
    }

    return v1 || v2 || v3

};

function checkSubtree(root:TreeNode, subRoot:TreeNode):boolean{
    if (root === null && subRoot === null){
        return true 
    }
    else if (root === null || subRoot === null){
        return false
    }
    else {
        if (root.val === subRoot.val){
            let v1 = checkSubtree(root.left, subRoot.left)
            let v2 = checkSubtree(root.right, subRoot.right)
            return v1 && v2
        }
        return false 
    }
}

File: src/my_project/medium_problems/from51to100/subtree_of_another_tree_logn.py
from typing import List 

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    def checkSubtree(self,root,subRoot): # function to check if tree and  subtree are equal

        if root is None and subRoot is None:
            return True 
        elif root is None or subRoot is None:
            return False 
        else: 
            if root.val == subRoot.val:
                v1 = self.checkSubtree(root.left, subRoot.left)
                v2 = self.checkSubtree(root.right, subRoot.right)
                
                return v1 and v2 
            return False 
        
    def isSubtree(self, root, subRoot) -> bool: # function to traverse the whole tree
        if root is None:
            return False 
        else: 
            v1 = False 
            if root.val ==subRoot.val:
                v1 = self.checkSubtree(root,subRoot)
                
            v2 = self.isSubtree(root.left, subRoot)
            v3 = self.isSubtree(root.right, subRoot)
            
        return v1 or v2 or v3
                